libraries/persistent_store/src/store.rs

```diff
@@ -60,4 +62,55 @@ impl From<StorageError> for StoreError {
     }
 }
 
+/// Result of store operations.
 pub type StoreResult<T> = Result<T, StoreError>;
+
+/// Progression ratio for store metrics.
+///
+/// This is used for the [capacity] and [lifetime] metrics. Those metrics are measured in words.
+///
+/// [capacity]: struct.Store.html#method.capacity
+/// [lifetime]: struct.Store.html#method.lifetime
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct StoreRatio {
+    /// How much of the metric is used.
+    pub used: usize,
+
+    /// How much of the metric can be used at most.
+    pub total: usize,
+}
+
+impl StoreRatio {
+    /// How much of the metric is remaining.
+    pub fn remaining(self) -> usize {
+        self.total - self.used
```

Comment:
What happens when `used` > `total`? It supposedly doesn't happen, but would be interesting to catch?

Reply:
Oops, I've made the fields public such that they can be accessed by the model, but the user should not be able to build such object. I fixed the visibility of the fields and added accessors such that the user cannot shot themselves in the foot. I also documented the invariant that `used <= total`.

Those kind of bugs are already caught by tests and fuzzing (both run in debug mode), which also catch some logical bugs too by regularly checking some internal invariants that are not always caught on Rust-level operations like here.

On a related note to arithmetic overflow, I'm looking at a possible intermediate PR to clarify the different integer types used in the crate (what is a natural number, what is a sequence of bits, etc) because you had a related question before (why use `u32` somewhere and `usize` otherwhere). In particular it would improve fuzzing because `usize` is variable-sized which is an issue because fuzzing is usually done with `usize = u64` while we are interested in `usize = u32`. This would catch additional bugs and seems cleaner because the behavior would depend less on the target architecture.

Reply:
Sounds good concerning arithmetic overflow. I was curious about your choice of `-` instead of i.e. a `checked_sub`.

Reply:
Yes, it's because the subtraction never fails (so `checked_sub` would introduce dead-code for the `None` branch), but turns out the question was useful to see that there was a visibility problem :-)